tests/gradients/test_param_shift_hessian.py

```diff
@@ -0,0 +1,453 @@
+
+        print("\n", hessian_qruns, "<", jacobian_qruns, "?")
+        print("\n", hessian_qruns, "<=", 2 ** 2 * 1, "?")
+        print("\n", hessian_qruns, "<=", 3 ** 1, "?")
+
+        assert np.allclose(hessian, jacobian)
+        assert hessian_qruns < jacobian_qruns
+        assert hessian_qruns <= 2 ** 2 * 1  # 1 = (1+2-1)C(2)
+        assert hessian_qruns <= 3 ** 1
+
+    def test_less_quantum_invocations2(self):
+        """Test that the hessian invokes less hardware executions than double differentiation"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, diff_method="parameter-shift", max_diff=2)
+        def circuit(x):
+            qml.RX(x[0], wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.RY(x[1], wires=0)
+            return qml.expval(qml.PauliZ(1))
+
+        x = np.array([0.1, 0.2], requires_grad=True)
+
+        with qml.Tracker(dev) as tracker:
+            hessian = qml.gradients.param_shift_hessian(circuit)(x)
+            hessian_qruns = tracker.totals["executions"]
+            jacobian = qml.jacobian(qml.jacobian(circuit))(x)
+            jacobian_qruns = tracker.totals["executions"] - hessian_qruns
+
+        print("\n", hessian_qruns, "<", jacobian_qruns, "?")
+        print("\n", hessian_qruns, "<=", 2 ** 2 * 3, "?")
+        print("\n", hessian_qruns, "<=", 3 ** 2, "?")
+
+        assert np.allclose(hessian, jacobian)
+        assert hessian_qruns < jacobian_qruns
+        assert hessian_qruns <= 2 ** 2 * 3  # 3 = (2+2-1)C(2)
+        assert hessian_qruns <= 3 ** 2
+
+    def test_less_quantum_invocations3(self):
+        """Test that the hessian invokes less hardware executions than double differentiation"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, diff_method="parameter-shift", max_diff=2)
+        def circuit(x):
+            qml.RX(x[0], wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.RY(x[1], wires=0)
+            qml.RZ(x[2], wires=1)
+            return qml.probs(wires=[0, 1])
+
+        x = np.array([0.1, 0.2, 0.3], requires_grad=True)
+
+        with qml.Tracker(dev) as tracker:
+            hessian = qml.gradients.param_shift_hessian(circuit)(x)
+            hessian_qruns = tracker.totals["executions"]
+            jacobian = qml.jacobian(qml.jacobian(circuit))(x)
+            jacobian_qruns = tracker.totals["executions"] - hessian_qruns
+
+        print("\n", hessian_qruns, "<", jacobian_qruns, "?")
+        print("\n", hessian_qruns, "<=", 2 ** 2 * 6, "?")
+        print("\n", hessian_qruns, "<=", 3 ** 3, "?")
+
+        assert np.allclose(hessian, jacobian)
+        assert hessian_qruns < jacobian_qruns
+        assert hessian_qruns <= 2 ** 2 * 6  # 6 = (3+2-1)C(2)
+        assert hessian_qruns <= 3 ** 3
+
+    def test_hessian_is_differentiable(self):
+        """Test that the 3rd derivate can be calculated via auto-differentiation"""
```

Comment:
😍 

I recommend also writing a test to ensure this works for Torch, TF, and JAX

Reply:
That's a good point, I will look into how to use those frameworks :)